src/osd/PG.cc

```diff
@@ -2789,10 +2797,38 @@ int PG::_prepare_write_info(map<string,bufferlist> *km,
 			    bool dirty_big_info,
 			    bool dirty_epoch)
 {
+  if (dirty_epoch) {
+    ::encode(epoch, (*km)[epoch_key]);
+  }
+
+  // try to do info efficiently?
+  if (!dirty_big_info) {
+    pg_fast_info_t fast;
+    fast.populate_from(info);
+    fast.apply_to(&last_written_info);
+    if (info == last_written_info) {
+      ::encode(fast, (*km)[fastinfo_key]);
+      return 0;
+    }
+    generic_derr << __func__ << " fastinfo failed, info:\n";
+    {
+      JSONFormatter jf(true);
+      jf.dump_object("info", info);
+      jf.flush(*_dout);
+    }
+    {
+      *_dout << "\nlast_written_info:\n";
+      JSONFormatter jf(true);
+      jf.dump_object("last_written_info", last_written_info);
+      jf.flush(*_dout);
+    }
+    *_dout << dendl;
+  }
```

Comment:
@liewegas My log is flooding with fastinfo failed error , here is one snippet..

2016-10-04 17:35:50.221450 7fb52ab9c700 -1 _prepare_write_info fastinfo failed, info:
{
    "pgid": "1.2b0",
    "last_update": "18'41",
    "last_complete": "18'41",
    "log_tail": "0'0",
    "last_user_version": 41,
    "last_backfill": "MAX",
    "last_backfill_bitwise": 0,
    "purged_snaps": "[]",
    "history": {
        "epoch_created": 15,
        "last_epoch_started": 16,
        "last_epoch_clean": 16,
        "last_epoch_split": 0,
        "last_epoch_marked_full": 0,
        "same_up_since": 15,
        "same_interval_since": 15,
        "same_primary_since": 15,
        "last_scrub": "0'0",
        "last_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_deep_scrub": "0'0",
        "last_deep_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_clean_scrub_stamp": "2016-10-04 17:34:54.968066"
    },
    "stats": {
        "version": "18'40",
        "reported_seq": "15",
        "reported_epoch": "18",
        "state": "active+clean",
        "last_fresh": "2016-10-04 17:35:42.281977",
        "last_change": "2016-10-04 17:34:56.004604",
        "last_active": "2016-10-04 17:35:42.281977",
        "last_peered": "2016-10-04 17:35:42.281977",
        "last_clean": "2016-10-04 17:35:42.281977",
        "last_became_active": "2016-10-04 17:34:56.002579",
        "last_became_peered": "2016-10-04 17:34:56.002579",
        "last_unstale": "2016-10-04 17:35:42.281977",
        "last_undegraded": "2016-10-04 17:35:42.281977",
        "last_fullsized": "2016-10-04 17:35:42.281977",
        "mapping_epoch": 15,
        "log_start": "0'0",
        "ondisk_log_start": "0'0",
        "created": 15,
        "last_epoch_clean": 16,
        "parent": "0.0",
        "parent_split_bits": 0,
        "last_scrub": "0'0",
        "last_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_deep_scrub": "0'0",
        "last_deep_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_clean_scrub_stamp": "2016-10-04 17:34:54.968066",
        "log_size": 40,
        "ondisk_log_size": 40,
        "stats_invalid": false,
        "dirty_stats_invalid": false,
        "omap_stats_invalid": false,
        "hitset_stats_invalid": false,
        "hitset_bytes_stats_invalid": false,
        "pin_stats_invalid": false,
        "stat_sum": {
            "num_bytes": 42991616,
            "num_objects": 11,
            "num_object_clones": 0,
            "num_object_copies": 20,
            "num_objects_missing_on_primary": 0,
            "num_objects_missing": 0,
            "num_objects_degraded": 0,
            "num_objects_misplaced": 0,
            "num_objects_unfound": 0,
            "num_objects_dirty": 11,
            "num_whiteouts": 0,
            "num_read": 0,
            "num_read_kb": 0,
            "num_write": 82,
            "num_write_kb": 41984,
            "num_scrub_errors": 0,
            "num_shallow_scrub_errors": 0,
            "num_deep_scrub_errors": 0,
            "num_objects_recovered": 0,
            "num_bytes_recovered": 0,
            "num_keys_recovered": 0,
            "num_objects_omap": 0,
            "num_objects_hit_set_archive": 0,
            "num_bytes_hit_set_archive": 0,
            "num_flush": 0,
            "num_flush_kb": 0,
            "num_evict": 0,
            "num_evict_kb": 0,
            "num_promote": 0,
            "num_flush_mode_high": 0,
            "num_flush_mode_low": 0,
            "num_evict_mode_some": 0,
            "num_evict_mode_full": 0,
            "num_objects_pinned": 0
        },
        "up": [
            11,
            0
        ],
        "acting": [
            11,
            0
        ],
        "blocked_by": [],
        "up_primary": 11,
        "acting_primary": 11
    },
    "empty": 0,
    "dne": 0,
    "incomplete": 0,
    "last_epoch_started": 16,
    "hit_set_history": {
        "current_last_update": "0'0",
        "history": []
    }
}

last_written_info:
{
    "pgid": "1.2b0",
    "last_update": "18'41",
    "last_complete": "18'41",
    "log_tail": "0'0",
    "last_user_version": 41,
    "last_backfill": "MAX",
    "last_backfill_bitwise": 0,
    "purged_snaps": "[]",
    "history": {
        "epoch_created": 15,
        "last_epoch_started": 16,
        "last_epoch_clean": 16,
        "last_epoch_split": 0,
        "last_epoch_marked_full": 0,
        "same_up_since": 15,
        "same_interval_since": 15,
        "same_primary_since": 15,
        "last_scrub": "0'0",
        "last_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_deep_scrub": "0'0",
        "last_deep_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_clean_scrub_stamp": "2016-10-04 17:34:54.968066"
    },
    "stats": {
        "version": "18'40",
        "reported_seq": "15",
        "reported_epoch": "18",
        "state": "active+clean",
        "last_fresh": "2016-10-04 17:35:42.281977",
        "last_change": "2016-10-04 17:34:56.004604",
        "last_active": "2016-10-04 17:35:42.281977",
        "last_active": "2016-10-04 17:35:42.281977",
        "last_peered": "2016-10-04 17:35:42.281977",
        "last_clean": "2016-10-04 17:35:42.281977",
        "last_became_active": "2016-10-04 17:34:56.002579",
        "last_became_peered": "2016-10-04 17:34:56.002579",
        "last_unstale": "2016-10-04 17:35:42.281977",
        "last_undegraded": "2016-10-04 17:35:42.281977",
        "last_fullsized": "2016-10-04 17:35:42.281977",
        "mapping_epoch": 15,
        "log_start": "0'0",
        "ondisk_log_start": "0'0",
        "created": 15,
        "last_epoch_clean": 16,
        "parent": "0.0",
        "parent_split_bits": 0,
        "last_scrub": "0'0",
        "last_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_deep_scrub": "0'0",
        "last_deep_scrub_stamp": "2016-10-04 17:34:54.968066",
        "last_clean_scrub_stamp": "2016-10-04 17:34:54.968066",
        "log_size": 40,
        "ondisk_log_size": 40,
        "stats_invalid": false,
        "dirty_stats_invalid": false,
        "omap_stats_invalid": false,
        "hitset_stats_invalid": false,
        "hitset_bytes_stats_invalid": false,
        "pin_stats_invalid": false,
        "stat_sum": {
            "num_bytes": 42991616,
            "num_objects": 11,
            "num_object_clones": 0,
            "num_object_copies": 20,
            "num_objects_missing_on_primary": 0,
            "num_objects_missing": 0,
            "num_objects_degraded": 0,
            "num_objects_misplaced": 0,
            "num_objects_unfound": 0,
            "num_objects_dirty": 10,
            "num_whiteouts": 0,
            "num_read": 0,
            "num_read_kb": 0,
            "num_write": 82,
            "num_write_kb": 41984,
            "num_scrub_errors": 0,
            "num_shallow_scrub_errors": 0,
            "num_deep_scrub_errors": 0,
            "num_objects_recovered": 0,
            "num_bytes_recovered": 0,
            "num_keys_recovered": 0,
            "num_objects_omap": 0,
            "num_objects_hit_set_archive": 0,
            "num_bytes_hit_set_archive": 0,
            "num_flush": 0,
            "num_flush_kb": 0,
            "num_evict": 0,
            "num_evict_kb": 0,
            "num_promote": 0,
            "num_flush_mode_high": 0,
            "num_flush_mode_low": 0,
            "num_evict_mode_some": 0,
            "num_evict_mode_full": 0,
            "num_objects_pinned": 0
        },
        "up": [
            11,
            0
        ],
        "acting": [
            11,
            0
        ],
        "blocked_by": [],
        "up_primary": 11,
        "acting_primary": 11
    },
    "empty": 0,
    "dne": 0,
    "incomplete": 0,
    "last_epoch_started": 16,
    "hit_set_history": {
        "current_last_update": "0'0",
        "history": []
    }
}

But, I am seeing pg_info is 178 bytes now , what's wrong ?